Add E2E tests GH workflow.

tomalec opened the pull request:
### Changes proposed in this Pull Request:

 To prevent issues like https://github.com/woocommerce/google-listings-and-ads/issues/1129



### Screenshots:

<!--- Optional --->


### Detailed test instructions:

1. See GH actions for this PR


### Additional comments:
1. We may think of sharing the build file between jobs, to save time, also, to run in only when the package or WP/WC versions actually changed.


<!--
Optional.
Enter a summary of all changes in this Pull Request, which will be added to the changelog if accepted.
Each line should start with `(Fix|Add|Tweak|Update) - `, for example:
> Fix - I took care of something that wasn't working.
> Add - I added something new that's pretty cool.
> Tweak - I made a small change.
> Update - I made big changes to something that wasn't broken.

Leave the "Changelog entry" header in place completely empty, without any summary if no changelog entry is needed.
If you remove the "Changelog entry" header, the title of Pull Request will be used as the changelog entry.
-->
### Changelog entry

>

tomalec committed:
Add E2E tests GH workflow.

tomalec committed:
Try Ubuntu 18.04 for E2E to recreade WCA setup.

tomalec committed:
Upload screenshots for E2E for debugging.

tomalec commented:
For some reason, in GHA the E2E's are stuck on OBW 

See screenshots https://github.com/woocommerce/google-listings-and-ads/actions/runs/1536018262
![Clicks-on-the-Marketing-GLA-link-clicks-on-the-call-to-action-setup-button-to-go-to-the-Setup-MC-page-should-get-to-the-accounts-setup](https://user-images.githubusercontent.com/17435/144636669-a45d6b8e-56fa-4fd1-bdce-56150cf7fa61.png)

In the local environment, it does not happen.
Does anybody have an idea why?

tomalec committed:
Skip the OBW in getting started E2E.

tomalec commented:
I worked around the OBW in https://github.com/woocommerce/google-listings-and-ads/pull/1136/commits/6b264a9be2001d4ae14bd08be8470b254760f38b with forced navigation to `WP_ADMIN_DASHBOARD = 'http://localhost:8084/wp-admin/'`, but I'm not sure if that's sustainable.

eason9487 commented on .github/workflows/e2e-tests.yml:
```diff
@@ -0,0 +1,63 @@
+        env:
+            WC_E2E_SCREENSHOTS: 1
+        run: |
+          WP_ENV_TESTS_PORT=8085 npm run test:e2e
```
❔ 
The `WP_ENV_TESTS_PORT=8085` seems to be left over from debugging?

eason9487 reviewed (approved):
LGTM. Left a non-blocking comment.

---

> I worked around the OBW in https://github.com/woocommerce/google-listings-and-ads/pull/1136/commits/6b264a9be2001d4ae14bd08be8470b254760f38b with forced navigation to `WP_ADMIN_DASHBOARD = 'http://localhost:8084/wp-admin/'`, but I'm not sure if that's sustainable.

I couldn't replicate the problem in my local environment. After some research, I suspect that it's caused by these:
1. redirect to the admin home page of WC
    https://github.com/woocommerce/woocommerce-admin/blob/4db2a7a1994ea52b13b5303314093044fbc4e436/src/Features/Onboarding.php#L136-L170
2. And then redirecting to WC setup wizard
    https://github.com/woocommerce/woocommerce-admin/blob/b3f605fae8573cdfa03188ea06c37f30dcbf44b9/client/homescreen/index.tsx#L25-L31

Maybe it could be addressed by adding a `woocommerce_prevent_automatic_wizard_redirect` filter or setting the wp_options `woocommerce_onboarding_profile` at somewhere by [option hook](https://developer.wordpress.org/reference/hooks/update_option_option/). Or instead, run a wp-cli command on the docker container of e2e test:

```bash
wp option update woocommerce_onboarding_profile '{"skipped": true, "completed": true}' --format=json
```

tomalec committed:
Remove debugging leftover from `e2e-test.yml`.

Addresses https://github.com/woocommerce/google-listings-and-ads/pull/1136#discussion_r763846721

tomalec commented:
> I couldn't replicate the problem in my local environment. […]
Me neither, that's why I was not happy with the forced fix.

> Maybe it could be addressed […] run a wp-cli command on the docker container of e2e test:
> 
> ```bash
> wp option update woocommerce_onboarding_profile '{"skipped": true, "completed": true}' --format=json
> ```

Nice! I don't know how you found it, as I cannot find the docs for it, but look more direct, explicit and sustainable, that forced redirect I did.

tomalec committed:
Skip OBW for E2E in more epxplicit and direct way,

by a wp-cli command on the docker container.
Addresses https://github.com/woocommerce/google-listings-and-ads/pull/1136#pullrequestreview-825072720
Reverts 6b264a9be2001d4ae14bd08be8470b254760f38b

Co-authored-by: Eason Su <eason.su.tw@gmail.com>

eason9487 commented:
The forced navigation problem is so interesting that I can't stop thinking after work. 😂 

> I couldn't replicate the problem in my local environment. […]
Me neither, that's why I was not happy with the forced fix.

> Maybe it could be addressed […] run a wp-cli command on the docker container of e2e test

After more investigations, the above assumptions are generally correct, and the root cause of force navigation starts from the `set_transient( '_wc_activation_redirect', 1, 30 )` [here](https://github.com/woocommerce/woocommerce/blob/d12717211a7cb97b6eef426d48219c3cba804d30/plugins/woocommerce/includes/class-wc-install.php#L430), and then [this condition](https://github.com/woocommerce/woocommerce-admin/blob/4db2a7a1994ea52b13b5303314093044fbc4e436/src/Features/Onboarding.php#L144) would be `true`. 

It's hard to replicate the problem in local development environment is because usually the gap time between running `npm docker:up` and `npm test:e2e` is more than 30 seconds, so the `_wc_activation_redirect` transient has already expired. But running e2e on GHA is much faster than in our local development environment. 🤷‍♂️ 

I tried deleting the `_wc_activation_redirect` transient by https://github.com/woocommerce/google-listings-and-ads/commit/460bf9ee451569e22ce0692f9306b8a52312fd60 and it works - https://github.com/woocommerce/google-listings-and-ads/runs/4446562083#step:9:15. Or I guess another solution could be sleep over 30 seconds [here](https://github.com/woocommerce/google-listings-and-ads/blob/72669ef95ae6217b66f0c4e5b1359cf8c2252162/.github/workflows/e2e-tests.yml#L49). 😆 

PS. The debug command `wp transient get _wc_activation_redirect` was not printed out from docker container.

eason9487 commented on .github/workflows/e2e-tests.yml:
```diff
@@ -0,0 +1,63 @@
+        with:
+            access_token: ${{ github.token }}
+
+      # Needed for Ubuntu 20.04.
+      - name: Kill mono to free 8084
+        run: |
+          sudo killall mono
```
💅 
The port of e2e test site can be changed by [this config](https://github.com/woocommerce/google-listings-and-ads/blob/72669ef95ae6217b66f0c4e5b1359cf8c2252162/tests/e2e/config/default.json#L2).

tomalec replied:
Yes, but then we'd have to update it locally, in the docs, and I'm afraid with the state of the WC e2e tooling, it may have some deeper hidden dependencies elsewhere. So I'd rather adapt the OS instance, to play with the defaults :see_no_evil:

tomalec commented:
> It's hard to replicate the problem in local development environment is because usually the gap time between running `npm docker:up` and `npm test:e2e` is more than 30 seconds, so the `_wc_activation_redirect` transient has already expired. But running e2e on GHA is much faster than in our local development environment. 🤷‍♂️ 
> 
> I tried deleting the `_wc_activation_redirect` transient by https://github.com/woocommerce/google-listings-and-ads/commit/460bf9ee451569e22ce0692f9306b8a52312fd60 and it works - https://github.com/woocommerce/google-listings-and-ads/runs/4446562083#step:9:15. Or I guess another solution could be sleep over 30 seconds [here](https://github.com/woocommerce/google-listings-and-ads/blob/72669ef95ae6217b66f0c4e5b1359cf8c2252162/.github/workflows/e2e-tests.yml#L49). 😆 

:open_mouth:  Awesome, that you made it work :superhero: 

A bit hacky, but way better than my redirect.

> PS. The debug command `wp transient get _wc_activation_redirect` was not printed out from docker container.

Yep, E2E does not allow us to log anything easily. It's logged to the container.
https://github.com/woocommerce/woocommerce/issues/30687

tomalec committed:
Delete `_wc_activation_redirect` transient not to interrupt our tests.

As suggested in 460bf9ee451569e22ce0692f9306b8a52312fd60
https://github.com/woocommerce/google-listings-and-ads/pull/1136#issuecomment-988095669

Co-authored-by: Eason Su <eason.su.tw@gmail.com>

tomalec commented:
@eason9487 Do you think it's good to go now? Are you fine with `killall mono`?

eason9487 commented:
> @eason9487 Do you think it's good to go now?

Sure! Thanks for adding the GHA workflow for e2e tests! 🎉 🚀
